storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -1255,9 +1260,17 @@ static dberr_t fsp_free_page(fil_space_t *space, uint32_t offset, mtr_t *mtr)
 	switch (state) {
 	case XDES_FREE_FRAG:
 	case XDES_FULL_FRAG:
-		if (!xdes_is_free(descr, offset % FSP_EXTENT_SIZE)) {
+	{
+		bool is_free= xdes_is_free(descr, offset % FSP_EXTENT_SIZE);
+		if (!is_free) {
 			break;
+		} else if (ignore_freed_page) {
+			return err;
 		}
+	}
+		/* fall through */
+	case XDES_FREE:
+		if (ignore_freed_page) return err;
 		/* fall through */
 	default:
 		space->set_corrupted();
```

Comment:
Could we tweak the rarely invoked function `fil_space_t::set_corrupted()` instead of adding a parameter and complex logic to a rather commonly invoked operation?